Model/Service/Index.php

```diff
@@ -0,0 +1,286 @@
+            $this->indexRepository->save($indexedProduct);
+            return $indexedProduct;
+        } catch (\Exception $e) {
+            $this->getLogger()->exception($e);
+            return null;
+        }
+    }
+
+    /**
+     * @param NostoIndexCollection $collection
+     * @param Store $store
+     * @throws NostoException
+     * @throws MemoryOutOfBoundsException
+     */
+    public function rebuildDirtyProducts(NostoIndexCollection $collection, Store $store)
+    {
+        $this->startBenchmark(
+            self::BENCHMARK_NAME_REBUILD,
+            self::BENCHMARK_BREAKPOINT_REBUILD
+        );
+        $collection->setPageSize(self::PRODUCT_DATA_BATCH_SIZE);
+        $iterator = new Iterator($collection);
+        $iterator->each(function (NostoProductIndex $item) {
+            if ($item->getIsDirty() === NostoProductIndex::DB_VALUE_BOOLEAN_TRUE) {
+                $this->rebuildDirtyProduct($item);
+                $this->tickBenchmark(self::BENCHMARK_NAME_REBUILD);
+            }
+            $this->checkMemoryConsumption('product rebuild');
+        });
+        $this->logBenchmarkSummary(self::BENCHMARK_NAME_REBUILD, $store);
+        $this->nostoSyncService->syncIndexedProducts($collection, $store);
+    }
+
+    /**
+     * Rebuilds a dirty indexed product data & defines it as out of sync
+     * if Nosto product data changed
+     *
+     * @param ProductIndexInterface $productIndex
+     * @return ProductIndexInterface|null
+     */
+    public function rebuildDirtyProduct(ProductIndexInterface $productIndex)
+    {
+        try {
+            /* @var Product $magentoProduct */
+            $magentoProduct = $this->loadMagentoProduct(
+                $productIndex->getProductId(),
+                $productIndex->getStoreId()
+            );
+            $store = $this->nostoHelperScope->getStore($productIndex->getStoreId());
+            $nostoProduct = $this->nostoProductBuilder->build($magentoProduct, $store);
+            $nostoIndexedProduct = $productIndex->getNostoProduct();
+            if ($nostoIndexedProduct instanceof NostoProductInterface === false ||
+                (
+                    $nostoProduct instanceof NostoProductInterface
+                    && !ProductUtil::isEqual($nostoProduct, $nostoIndexedProduct)
+                )
+            ) {
+                $productIndex->setNostoProduct($nostoProduct);
+                $productIndex->setInSync(false);
+            }
+            $productIndex->setIsDirty(false);
+            $this->indexRepository->save($productIndex);
+            return $productIndex;
+        } catch (\Exception $e) {
+            $this->getLogger()->exception($e);
+            return null;
+        }
+    }
+
+    /**
+     * @param ProductCollection $collection
+     * @param array $ids
+     * @param Store $store
+     * @throws NostoException
+     */
+    public function markProductsAsDeletedByDiff(ProductCollection $collection, array $ids, Store $store)
```

Comment:
What is this method doing? It is very very cryptic.

Reply:
It's cryptic, I agree. This is a shortcut method for marking cached Nosto products as deleted. Basically it checks what product ids from `$ids` are not found from the collection and marks those ids as deleted.